src/field_theory/algebraic_closure.lean

```diff
@@ -0,0 +1,283 @@
+is_alg_closed.of_exists_root _ $ λ p _ hp, complex.exists_root $ degree_pos_of_irreducible hp
+
+/-- Typeclass for an extension being an algebraic closure. -/
+@[class] def is_alg_closure (K : Type v) [field K] [algebra k K] : Prop :=
+is_alg_closed K ∧ algebra.is_algebraic k K
+
+namespace algebraic_closure
+
+open mv_polynomial
+
+/-- The subtype of monic irreducible polynomials -/
+@[reducible] def monic_irreducible : Type u :=
+{ f : polynomial k // monic f ∧ irreducible f }
+
+/-- Sends a monic irreducible polynomial `f` to `f(x_f)` where `x_f` is a formal indeterminate. -/
+def eval_X_self (p : monic_irreducible k) : mv_polynomial (monic_irreducible k) k :=
+polynomial.eval₂ mv_polynomial.C (X p) p
+
+/-- The span of `f(x_f)` across monic irreducible polynomials `f` where `x_f` is an indeterminate. -/
+def span_eval : ideal (mv_polynomial (monic_irreducible k) k) :=
+ideal.span $ set.range $ eval_X_self k
+
+/-- Given a finset of monic irreducible polynomials, construct an algebra homomorphism to the
+splitting field of the product of the polynomials sending each indeterminate `x_f` represented by
+the polynomial `f` in the finset to a root of `f`. -/
+def to_splitting_field (s : finset (monic_irreducible k)) :
+  mv_polynomial (monic_irreducible k) k →ₐ[k] splitting_field (∏ x in s, x : polynomial k) :=
+mv_polynomial.aeval $ λ f, if hf : f ∈ s
+then root_of_splits _ ((splits_prod_iff _ $ λ (j : monic_irreducible k) _, j.2.2.ne_zero).1
+    (splitting_field.splits _) f hf)
+  (mt is_unit_iff_degree_eq_zero.2 f.2.2.not_unit)
+else 37
+
+theorem to_splitting_field_eval_X_self {s : finset (monic_irreducible k)} {f} (hf : f ∈ s) :
+  to_splitting_field k s (eval_X_self k f) = 0 :=
+by { rw [to_splitting_field, eval_X_self, ← alg_hom.coe_to_ring_hom, hom_eval₂,
+         alg_hom.coe_to_ring_hom, mv_polynomial.aeval_X, dif_pos hf,
+         ← algebra_map_eq, alg_hom.comp_algebra_map],
+  exact map_root_of_splits _ _ _ }
+
+theorem span_eval_ne_top : span_eval k ≠ ⊤ :=
+begin
+  rw [ideal.ne_top_iff_one, span_eval, ideal.span, ← set.image_univ, finsupp.mem_span_iff_total],
+  rintros ⟨v, _, hv⟩, replace hv := congr_arg (to_splitting_field k v.support) hv,
+  rw [alg_hom.map_one, finsupp.total_apply, finsupp.sum, alg_hom.map_sum, finset.sum_eq_zero] at hv,
+  { exact zero_ne_one hv },
+  intros j hj, rw [smul_eq_mul, alg_hom.map_mul, to_splitting_field_eval_X_self k hj, mul_zero]
+end
+
+/-- A random maximal ideal that contains `span_eval k` -/
+def max_ideal : ideal (mv_polynomial (monic_irreducible k) k) :=
+classical.some $ ideal.exists_le_maximal _ $ span_eval_ne_top k
+
+instance max_ideal.is_maximal : (max_ideal k).is_maximal :=
+(classical.some_spec $ ideal.exists_le_maximal _ $ span_eval_ne_top k).1
+
+theorem le_max_ideal : span_eval k ≤ max_ideal k :=
+(classical.some_spec $ ideal.exists_le_maximal _ $ span_eval_ne_top k).2
+
+/-- The first step: adjoin a root of all monic polynomials -/
```

Comment:
Since the construction of `algebraic_closure` is split over a few definitions, could you add a short section `## Implementation notes` to the module doc?
```suggestion
/-- The first step of constructing `algebraic_closure`: adjoin a root of all monic polynomials -/
```